### PR TITLE
Update C++ UTF8 code to use std::byte

### LINKED ISSUE #1930: We should fix the UTF8-related code in a follow-up PR, and remove some related reinterpret_cast.
We should fix the UTF8-related code in a follow-up PR, and remove some related reinterpret_cast.

_Originally posted by @bernardnormier in https://github.com/zeroc-ice/ice/pull/1928#pullrequestreview-1934590815_

### PR DESCRIPTION
**Pull request by externl**
Closes #1930

### CODE REVIEW

File: cpp/include/Ice/IconvStringConverter.h [hunk 1]
```diff
@@ -236,15 +236,15 @@ namespace IceInternal
         //
         do
         {
-            size_t bytesused = 0;
-            if (outbuf != 0)
+            size_t bytesUsed = 0;
+            if (outbuf != nullptr)
```
Comment (pepone) [inline cpp/include/Ice/IconvStringConverter.h:240]:
```suggestion
            if (outbuf)
```

File: cpp/src/Ice/OutputStream.cpp [hunk 2]
```diff
@@ -24,21 +24,21 @@ using namespace IceInternal;
...
     public:
         StreamUTF8BufferI(OutputStream& stream) : _stream(stream) {}
 
-        uint8_t* getMoreBytes(size_t howMany, uint8_t* firstUnused)
+        byte* getMoreBytes(size_t howMany, byte* firstUnused) final
         {
             assert(howMany > 0);
 
-            if (firstUnused != 0)
+            if (firstUnused != nullptr)
```
Comment (pepone) [inline cpp/src/Ice/OutputStream.cpp:36]:
```suggestion
            if (firstUnused)
```

File: cpp/src/IceUtil/StringConverter.cpp [hunk 3]
```diff
@@ -179,32 +179,32 @@ namespace
...
     public:
         //
         // Returns the first unused byte in the resized buffer
         //
-        uint8_t* getMoreBytes(size_t howMany, uint8_t* firstUnused)
+        byte* getMoreBytes(size_t howMany, byte* firstUnused) final
         {
             size_t bytesUsed = 0;
-            if (firstUnused != 0)
+            if (firstUnused != nullptr)
```
Comment (pepone) [inline cpp/src/IceUtil/StringConverter.cpp:191]:
```suggestion
            if (firstUnused)
```

**Review by pepone — APPROVED** [review]

File: cpp/src/IceUtil/StringConverter.cpp [hunk 4]
```diff
@@ -47,7 +47,7 @@ namespace
     class UnicodeWstringConverter : public WstringConverter
     {
     public:
-        virtual uint8_t* toUTF8(const wchar_t* sourceStart, const wchar_t* sourceEnd, UTF8Buffer& buffer) const
+        virtual byte* toUTF8(const wchar_t* sourceStart, const wchar_t* sourceEnd, UTF8Buffer& buffer) const
```
Comment (bernardnormier) [inline cpp/src/IceUtil/StringConverter.cpp:50]: virtual -> final. Also make the class final.

**Review by bernardnormier — APPROVED** [review]

**Pull request merged**